lib/plugins/amplitudePlugin.ts

```diff
@@ -0,0 +1,120 @@
+import { add, init, track } from '@amplitude/analytics-node'
+import type {
+  AmplitudeReturn,
+  BaseEvent,
+  NodeOptions,
+  Plugin,
+  Result,
+} from '@amplitude/analytics-types'
+import type {
+  FastifyInstance,
+  FastifyReply,
+  FastifyRequest,
+  HookHandlerDoneFunction,
+} from 'fastify'
+import fp from 'fastify-plugin'
+
+async function plugin(
+  fastify: FastifyInstance,
+  config: AmplitudeConfig,
+  next: (err?: Error) => void,
+) {
+  if (!config.isEnabled) {
+    return next()
+  }
+
+  if (!config.apiKey) {
+    throw Error('Amplitude key not defined')
+  }
+
+  await init(config.apiKey, config.options).promise
+
+  if (config.apiUsageTracking) {
+    enableApiUsageTracking(fastify, config.apiUsageTracking)
+  }
+  if (config.plugins) {
+    // @ts-expect-error
+    config.plugins.forEach((e) => add(e))
+  }
+
+  return next()
+}
+
+function enableApiUsageTracking(
+  fastify: FastifyInstance,
+  eventCreationFn: createApiTrackingEventFn,
+) {
+  fastify.addHook(
+    'onResponse',
+    (req: FastifyRequest, res: FastifyReply, done: HookHandlerDoneFunction) => {
+      const event = eventCreationFn(req, res)
+      if (event) {
+        amplitudeTrack(event)
+      }
+      return done()
+    },
+  )
+}
+
+/**
+ * Callback used to create the events that will be used to automatically
+ * track the API usage
+ */
+export type createApiTrackingEventFn = (req: FastifyRequest, res: FastifyReply) => BaseEvent | null
+
+/**
+ * Configuration to set up the Amplitude Plugin.
+ *
+ * @property {boolean} isEnabled Flag to enable or disable the plugin.
+ *
+ * @property {string} apiKey Amplitude api key, please get it from your
+ * Amplitude project
+ *
+ * @property {NodeOptions} Amplitude configuration, please check
+ * [this](https://amplitude.github.io/Amplitude-TypeScript/modules/_amplitude_analytics_node.Types.html#NodeOptions)
+ * to learn more.
+ *
+ * @property {createApiTrackingEventFn} apiUsageTracking Callback used to create
+ * the event that will be send automatically to track the API usage. If not
+ * specified the API usage track will be disabled.
+ *
+ * @property {Plugin[]} plugins Allow to extend plugin behavior by, for example,
+ * modifying event properties. Please check
+ * [this](https://www.docs.developers.amplitude.com/data/sdks/typescript-node/#plugins)
+ * to see how it works
+ */
+export interface AmplitudeConfig {
+  isEnabled: boolean
+  apiKey?: string
+  options?: NodeOptions
+  apiUsageTracking?: createApiTrackingEventFn
+  plugins?: Plugin[]
+}
+
+/**
+ * Use this method to register the amplitude plugin on your fastify instance.
```

Comment:
I would put this in README instead

Reply:
I thought about putting it there, but we didn't do it with the other plugins we have implemented so I went with this approach.

If you think that adding some doc on the readme is valuable I will be happy to do it, just mentioning it to make sure it is the way to go (I don't know why the other plugins haven't got doc)

Reply:
we don't have great documentation in this module right now, unfortunately; it would be great to document new functionality from the get go

we've started documenting what we have, e. g. see https://github.com/lokalise/node-core

You may want to sync with Daria, as she is currently documenting existing code (including fastify-extras), so that there is a consistent place and structure for the docs

Reply:
I see, and I agree with you :) 

I have discussed with Daria regarding the `library`, and there is already a draft PR that includes some documentation. However, Daria will be on vacation starting tomorrow. To ensure a smooth process and avoid any conflicts with Daria's PR, I suggest we release the PR without the documentation as soon as possible. This way, we can proceed with the next amplitude task. Subsequently, I can open a new PR that links to Daria's branch and includes the plugin documentation. 
Does it sound good? please feel free to disagree.

Reply:
sure, sounds like a plan!